pkg/api/conversions/conversions.go

```diff
@@ -0,0 +1,103 @@
+	}
+	if err = json.Unmarshal(inStatusJSON, &out.Status); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal downstreammachine status: %v", err)
+	}
+
+	out.Spec.ObjectMeta = in.Spec.ObjectMeta
+	out.Spec.Taints = in.Spec.Taints
+
+	providerConfigRaw, err := json.Marshal(in.Spec.ProviderConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	out.Spec.ProviderConfig = clusterv1alpha1.ProviderConfig{Value: &runtime.RawExtension{Raw: providerConfigRaw}}
+
+	for _, inRole := range in.Spec.Roles {
+		if inRole == machinev1alpha1downstream.MasterRole {
+			out.Spec.Roles = append(out.Spec.Roles, clustercommon.MasterRole)
+		}
+		if inRole == machinev1alpha1downstream.NodeRole {
+			out.Spec.Roles = append(out.Spec.Roles, clustercommon.NodeRole)
+		}
+	}
+
+	out.Finalizers = in.Finalizers
+
+	// This currently results in in.Spec.Versions.ContainerRuntime being dropped,
+	// because it was removed from the upstream type in
+	// https://github.com/kubernetes-sigs/cluster-api/pull/240
+	// To work around this, we put it into the providerConfig
+	inMachineVersionJSON, err := json.Marshal(in.Spec.Versions)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal downstreammachine version: %v", err)
+	}
+	if err = json.Unmarshal(inMachineVersionJSON, &out.Spec.Versions); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal downstreammachine version: %v", err)
+	}
+
+	providerConfigMap, err := addContainerRuntimeInfoToProviderConfig(*out.Spec.ProviderConfig.Value,
+		in.Spec.Versions.ContainerRuntime)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add containerRuntimeInfo to providerConfig: %v", err)
+	}
+	out.Spec.ProviderConfig.Value.Raw, err = json.Marshal(providerConfigMap)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshall providerconfig after adding containerRuntimeInfo: %v", err)
+	}
+
+	out.Spec.ConfigSource = in.Spec.ConfigSource
+
+	return out, err
+}
+
+func addContainerRuntimeInfoToProviderConfig(providerConfigValue runtime.RawExtension, containerRuntimeInfo machinev1alpha1downstream.ContainerRuntimeInfo) (map[string]interface{}, error) {
+	providerConfigMap := map[string]interface{}{}
+	if err := json.Unmarshal(providerConfigValue.Raw, &providerConfigMap); err != nil {
+		return nil, fmt.Errorf("failed to unmarshall provider config into map: %v", err)
+	}
+	// The JSON unmarshall makes the map a null pointer if providerConfigValue.Raw is empty
+	if providerConfigMap == nil {
+		providerConfigMap = map[string]interface{}{}
+	}
+	if val, ok := providerConfigMap["operatingSystemSpec"]; ok {
+		if valMap, ok := val.(map[string]interface{}); ok {
+			valMap["containerRuntimeInfo"] = containerRuntimeInfo
+			providerConfigMap["operatingSystemSpec"] = valMap
+			return providerConfigMap, nil
+		}
+	}
+	providerConfigMap["operatingSystemSpec"] = map[string]machinev1alpha1downstream.ContainerRuntimeInfo{"containerRuntimeInfo": containerRuntimeInfo}
```

Comment:
not sure if `containerRuntimeInfo` is part of `operatingSystemSpec`, how about making `containerRuntimeInfo` top level ?